IA-1659 V1 of entity API for mobile

madewulf opened the pull request:
Explain what problem this PR is resolving

## Self proof reading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] Did I add translations
- [x] My migrations file are included
- [x] Are there enough tests
## Changes

Explain the changes that were made. The idea is not to list exhaustively all the changes made (Github already provides a full diff), but to help the reviewers better understand:
- which specific file changes go together, e.g: when creating a table in the front-end, there usually is a config file that goes with it
- the reasoning behind some changes, e.g: deleted files because they are now redundant
- the behaviour to expect, e.g: tooltip has purple background color because the client likes it so, changed a key in the API response to be consistent with other endpoints


## How to test

Explain how to test your PR.
If a specific config is required explain it here (dataset, account, profile, ...)

## Print screen / video

Upload here print screens or videos showing the changes

## Notes

Things that the reviewers should know: known bugs that are out of the scope of the PR, other trade-offs that were made.
If the PR depends on a PR in bluesquare-components, or merges into another PR (i.o. main), it should also be mentioned here

xavierbarbosa reviewed (commented):
Looking good so far 👍

madewulf committed:
V1 of entity API for mobile

Commit:
Merge remote-tracking branch 'origin/main' into IA-1659-download-entities

Commit:
Modification requested by Benjamin

id should return the uuid (therefore, the field uuid can be removed)

attributes should return the uuid of the instance and be renamed into defining_instance_id

entity_type should be renamed into entity_type_id

created_at and updated_at should return Doubles (not formatted into String as they are right now)

In the instance serializer:

id should return the uuid (therefore, the field uuid can be removed)

form_id should return the value from Form.id not Form.form_id

org_unit should be renamed into org_unit_id

created_at and updated_at should return Doubles (not formatted into String as they are right now)

Commit:
modification on json response

Commit:
remove get field method

Commit:
add test

Commit:
add test

mathvdh commented:
Le message au début du commit n'apporte aucunes informations ?? On dirait que c'est le template non modifié ?

mathvdh reviewed (approved):
Cela me semble assez straightforward ... Je ne sais pas trop dans quel cadre ce truc est utilisé donc mon avis est simplement basé sur un survol du code

olethanh commented:
@bmonjoie can you confirm that the PR is what you expect in term of API?

bmonjoie commented on iaso/api/mobile/entity.py:
```diff
@@ -0,0 +1,106 @@
+
+    form_id = serializers.IntegerField(read_only=True, source="id")
+    id = serializers.CharField(read_only=True, source="uuid")
+    form_version_id = serializers.CharField(read_only=True, source="form.form.id")
```
That looks weird... Are you sure the source is the right one?

SimonDus replied:
I made the change, can you check if it's ok ?

bmonjoie replied:
It's not the right value, I'm looking for the version id, not the form's form_id property.
When fixed, don't forget to add the check in the test

bmonjoie commented on iaso/api/mobile/entity.py:
```diff
@@ -0,0 +1,106 @@
+    @staticmethod
+    def get_instances(entity: Entity):
+        res = []
+        for instance in entity.non_deleted_instances:  # type: ignore
```
If we give a date, it should be passed down here to limit the instances to the ones that have been updated after the given date as well, but we can do this optimization later.

SimonDus replied:
Limit date is in the get_queryset method not in the serializer...

bmonjoie replied:
My bad, I didn't see that they were already filtered

bmonjoie commented on iaso/api/mobile/entity.py:
```diff
@@ -0,0 +1,106 @@
+
+    details = /api/mobile/entity/id
+
+    sample usage: /api/mobile/entity/?limit_date=2022-12-29&limit=1&page=1
```
The endpoints should use plurals:
* /api/mobile/entities
* /api/mobile/entities/id

as previously agreed

bmonjoie commented on iaso/api/mobile/entity.py:
```diff
@@ -0,0 +1,106 @@
+
+    It's possible to filter out entities with no activity before a certain date with the parameter limit_date
+
+    details = /api/mobile/entity/id
```
I believe that the default is to use the "id" and not the "uuid", therefore, this endpoint is not correct. The "id" field is an implementation detail of the backend and should not be exposed.

bmonjoie commented on iaso/urls.py:
```diff
@@ -117,6 +118,7 @@
 router.register(r"tasks", TaskSourceViewSet, basename="tasks")
 router.register(r"comments", CommentViewSet, basename="comments")
 router.register(r"entity", EntityViewSet, basename="entity")
+router.register(r"mobile/entity", MobileEntityViewSet, basename="entity")
```
As stated, we agreed those should be plurals

bmonjoie commented on iaso/api/mobile/entity.py:
```diff
@@ -0,0 +1,106 @@
+        model = Instance
+        fields = ["id", "form_id", "form_version_id", "created_at", "updated_at", "org_unit_id", "json"]
+
+    form_id = serializers.IntegerField(read_only=True, source="id")
```
I might be wrong but this seems weird as well

SimonDus replied:
I made the change, can you check if it's ok ?

bmonjoie replied:
I believe this one is OK now, yet don't forget to add a test to check it is

Commit:
fix url from entity to entities - lookup_field from id to uuid

Commit:
Merge remote-tracking branch 'origin/develop' into IA-1659-download-entities

Commit:
get the correct form_version.id for form_version_id field in json response

Commit:
fix mypy

Commit:
move test entities to the tests/api folder and add a check to mobile entities tests

Commit:
remove test entities from old folder

Commit:
Merge remote-tracking branch 'origin/main' into IA-1659-download-entities

Commit:
black format

niconoe reviewed (approved):
Also seems quite straightforward to me even if I don't have the full context

bmonjoie commented on iaso/api/mobile/entity.py:
```diff
@@ -0,0 +1,115 @@
+        for instance in entity.non_deleted_instances:  # type: ignore
+            res.append(MobileEntityAttributesSerializer(instance).data)
+
+        return res
```
Just for curiosity, wouldn't it be better to use `map`?

```suggestion
    @staticmethod
    def get_instances(entity: Entity):
        return map(
            lambda instance: MobileEntityAttributesSerializer(instance).data, 
            entity.non_deleted_instances
        )
```

bmonjoie commented on iaso/api/mobile/entity.py:
```diff
@@ -0,0 +1,115 @@
+            version_id=obj.json.get("_version"), form__form_id=obj.form.form_id  # type: ignore
+        ).id
+
+        return form_version_id
```
It would probably make more sense to use `form.id` instead of `form.form_id` since `form.id` is the primary key. Also, why are you storing the result in a variable before returning it?

```suggestion
    def get_form_version_id(obj: Instance):
        return FormVersion.objects.get(
            version_id=obj.json.get("_version"), form__id=obj.form.id  # type: ignore
        ).id
```

bmonjoie commented on iaso/api/mobile/entity.py:
```diff
@@ -0,0 +1,115 @@
+
+    def get_queryset(self):
+        profile = self.request.user.iaso_profile
+        orgunits = OrgUnit.objects.hierarchy(profile.org_units.all())
```
What is the value of this if the user is not attached to any org units?

bmonjoie commented on iaso/tests/api/test_entities.py:
```diff
@@ -482,12 +482,67 @@ def test_handle_export_entity_type_empty_field_list(self):
+    def test_entity_mobile(self):
+        self.client.force_authenticate(self.yoda)
+
+        self.yoda.iaso_profile.org_units.set([self.jedi_council_corruscant])
```
Can you add a test when there are no org_units set in the profile?

Commit:
return all entities if there in no org_units in iaso_profile - small refactor - add tests

Commit:
mypy
